### ISSUE TITLE
[SPIKE] Proof of concept for removing log-and-throw and log-and-rethrow

### ISSUE THREAD

**Issue by DeltaMichael** [vmware/versatile-data-kit#2599]
## Overview

Before executing https://github.com/vmware/versatile-data-kit/issues/2570, we need to test the concept on some real scenarios and see the actual user experience. A good way to do this is to execute some failing example jobs in Supercollider with the change in vdk-core and check the logs.

## DoD
1. We have a proof-of-concept branch + short analysis and examples for removing the pattern
2. Sync conclusions with rest of team

**Comment by DeltaMichael** [comment]
It seems our error format doesn't work well when wrapping our custom errors in other custom errors. Example with added delimiters.

```
-----------------------------
An error in data job code  occurred. The error should be resolved by ResolvableByActual.USER. Here are the details:
  WHAT HAPPENED : Processing step 30_ingest_into_table.py completed with error.
WHY IT HAPPENED : An exception occurred, exception message was: -----------------------------
An error in data job code  occurred. The error should be resolved by ResolvableByActual.USER. Here are the details:
  WHAT HAPPENED : Provided method, SQLITE, has invalid value.
WHY IT HAPPENED : VDK was run with method=SQLITE, however SQLITE is not part of the available ingestion mechanisms.
   CONSEQUENCES : I'm rethrowing this exception to my caller to process.Most likely this will result in failure of current Data Job.
COUNTERMEASURES : Provide either valid value for method, or install ingestion plugin that supports this type. Currently possible values are ['http', 'file', 'sqlite']
-----------------------------
   CONSEQUENCES : I will not process the remaining steps (if any), and this Data Job execution will be marked as failed.
COUNTERMEASURES : See exception and fix the root cause, so that the exception does not appear anymore.
-----------------------------
```

It's extremely hard to understand what kind of error happened without the added delimiters.

**Comment by DeltaMichael** [comment]
It seems we conflate logging the an error with the actual exception error messages. We should have a few simple rules:

1. We should log error messages and throw an exception, but the message we log cannot be the error message of the exception. This should apply to logging and throwing in internals.
2. We should analyze the traceback and build the final exception message at the app exit point (top-level components like job_input), e.g. when we know we're going to throw the exception. At this point, if it's a User error, we can cherry-pick all the user and non-internal stack frames and log them, then throw the exception with the same message. If it's a platform error, we should log all relevant stack frames again (probably all stack frames) and throw the exception.

**Comment by DeltaMichael** [comment]
I was messing around with the code some more and ended up with this diff https://github.com/vmware/versatile-data-kit/pull/2634/files

These are the issues that popped up

1. If vdk exits with exit code 1, it prints out the stack trace for the error. The problem with this is that we already logged the stack trace when we handled the exception, so we end up outputting the same thing twice. Hacking the code to always return 0 shortens the stack trace by half. There should be a nicer way to do it, however.

2. We log too many exceptions along the call stack, so error messages from the exception we log in the `run` method end up  in the traceback logs.

3. Our exception format does not support nesting in the cases where one VKD exception concatenates the message from another VDK exception with its own error message.
